tools/tf2openapi/generator/generate_test.go

```diff
@@ -0,0 +1,308 @@
+								TensorShape: &framework.TensorShapeProto{
+									Dim: []*framework.TensorShapeProto_Dim{
+										{Size: -1},
+										{Size: 3},
+									},
+									UnknownRank: false,
+								},
+							},
+						},
+						Outputs: map[string]*pb.TensorInfo{
+							"outputTensorName": {
+								Dtype: framework.DataType_DT_INT8,
+								TensorShape: &framework.TensorShapeProto{
+									Dim: []*framework.TensorShapeProto_Dim{
+										{Size: -1},
+										{Size: 3},
+									},
+									UnknownRank: false,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	_, specErr := generator.GenerateOpenAPI(model)
+	expectedErr := fmt.Sprintf(types.UnsupportedDataTypeError, "inputTensorName", "DT_COMPLEX128")
+	g.Expect(specErr).To(gomega.MatchError(expectedErr))
+}
+
+func TestGenerateOpenAPISpecGenerationErr(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+	generator := defaultGenerator()
+	model := &pb.SavedModel{
+		MetaGraphs: []*pb.MetaGraphDef{
+			{
+				MetaInfoDef: &pb.MetaGraphDef_MetaInfoDef{
+					Tags: []string{
+						"serve",
+					},
+				},
+				SignatureDef: map[string]*pb.SignatureDef{
+					"serving_default": {
+						MethodName: "tensorflow/serving/classify",
+						Inputs: map[string]*pb.TensorInfo{
+							"inputTensorName": {
+								Dtype: framework.DataType_DT_INT8,
+								TensorShape: &framework.TensorShapeProto{
+									Dim: []*framework.TensorShapeProto_Dim{
+										{Size: -1},
+										{Size: 3},
+									},
+									UnknownRank: false,
+								},
+							},
+						},
+						Outputs: map[string]*pb.TensorInfo{
+							"outputTensorName": {
+								Dtype: framework.DataType_DT_INT8,
+								TensorShape: &framework.TensorShapeProto{
+									Dim: []*framework.TensorShapeProto_Dim{
+										{Size: -1},
+										{Size: 3},
+									},
+									UnknownRank: false,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	_, specErr := generator.GenerateOpenAPI(model)
+	expectedErr := fmt.Sprintf(SpecGenerationError, types.UnsupportedAPISchemaError)
+	g.Expect(specErr).To(gomega.MatchError(expectedErr))
+}
+
+
+func TestGenerateOpenAPIForRowFmtMultipleTensors(t *testing.T) {
+	// model src: gs://kfserving-samples/models/tensorflow/flowers
+	g := gomega.NewGomegaWithT(t)
+	model := model(t, "TestRowFmtMultipleTensors")
+	generator := defaultGenerator()
+	spec, specErr := generator.GenerateOpenAPI(model)
+	g.Expect(specErr).Should(gomega.BeNil())
+
+	swagger := &openapi3.Swagger{}
+	g.Expect(json.Unmarshal([]byte(spec), &swagger)).To(gomega.Succeed())
+
+	expectedSpec := string(openAPI(t, "TestRowFmtMultipleTensors"))
+	expectedSwagger := &openapi3.Swagger{}
+	// remove any formatting from expectedSpec
+	buffer := new(bytes.Buffer)
+	if err := json.Compact(buffer, []byte(expectedSpec)); err != nil {
+		t.Fatal(err)
+	}
+	g.Expect(json.Unmarshal(buffer.Bytes(), &expectedSwagger)).To(gomega.Succeed())
+
+	// test equality, ignoring order in JSON arrays
```

Comment:
Is this nondeterministic?

Reply:
Right now, the order of tensors names in the array is non-deterministic because I loop over a map to extract the tensors. I don't think it's a concern because the JSON array of tensor names constitutes the "required" properties, which can be in any order.

Would it be better to take a little bit of extra time to sort the tensors so they're always in alphabetical order, for example?

Reply:
Simply ignoring the ordering (but asserting set existance) of JSON arrays in the test is fine.